### ISSUE TITLE
Unnecessary jmp inserted with multiple compilation units

### ISSUE THREAD

**Issue by gurok** [pts/owccdos#4]
Consider this example:

main.c:
```
#include <dosmc.h>

void myoputcrlf(void);

void _start(void)
{
	myoputcrlf();
	myoputcrlf();

	return;
}

void myoputcrlf(void)
{
	oputcrlf();

	return;
}
```

Compiled with:
`./dosmc -mt -fo=/code/test.com /code/main.c`

It comes to 18 bytes:
`E8 00 00 52 B2 0D B4 02 CD 21 B2 0A B4 02 CD 21 5A C3`

With the following disassembly (courtesy of https://disasm.pro/):
```
call 3
push dx
mov dl, 0xd
mov ah, 2
int 0x21
mov dl, 0xa
mov ah, 2
int 0x21
pop dx
ret
```

So far, so good and pretty optimal for what it does. Now consider this:

main.c
```
#include <dosmc.h>

void myoputcrlf(void);

void _start(void)
{
	myoputcrlf();
	myoputcrlf();

	return;
}
```

other.c
```
void myoputcrlf(void)
{
	oputcrlf();

	return;
}
```

Compiled with:
`./dosmc -mt -fo=/code/test.com /code/main.c /code/other.c`

I get 21 bytes:
`E8 03 00 E9 00 00 52 B2 0D B4 02 CD 21 B2 0A B4 02 CD 21 5A C3`

Which, when disassembled, show a new (useless) `jmp` instruction just after the first call:
```
call 6
jmp 6
push dx
mov dl, 0xd
mov ah, 2
int 0x21
mov dl, 0xa
mov ah, 2
int 0x21
pop dx
ret
```

Can the compiler be made a little smarter in situations like this? Or am I just doing things wrong? I read your documentation on the project page and I believe I'm doing things the right way, i.e. `myoputcrlf` is predeclared and `other.c` is compiled after `main.c`.

**Comment by pts** [comment]
Thank you for the very detailed report! Based on the info you provided it was easy for me to figure out what was going on.

**TL;DR** You are using it correctly, and unfortunately you have to use a single .c source file to get the smallest output.

If *dosmc* was smarter, it could eliminate the unnecessary *jmp* instruction. Modern C compilers and linkers do similar optimizations, it's part of their link-time optimization. Unfortunately *dosmc* doesn't do it. It's not easy to add it in general, so it's unlikely to be added in the near future, unless there are volunteers.

Here is how it works: with a single .c file, the *wcc* compiler used by *dosmc* emits a single .obj file containing the code without the *jmp*, because *wcc* is smart enough to eliminate an unnecessary *jmp* in this case. With two .c files, the *wcc* compiler emits an .obj file containing _start(): a *call* and a *jmp* instruction, and also another .obj file containing myoutputcrlf(): the rest of the instructions, up to *ret*. The fundamental difference is that there are now two invocations of the *wcc* compiler, and it is not able to eliminate code across .obj files. The *dosmc* linker just concatenates code in .obj files, and adjusts some addresses (so that the call and jmp points to address 6), it doesn't add or remove remove any code bytes. It just sees bytes, it has no understanding of instructions, thus it has no chance to detect that a *jmp* instruction can be removed. This is a missed optimization opportunity in *dosmc*. To make it work, lots of code has to be written for *dosmc* to make it understand instructions and control flow.

 in the two-c-files case *dosmc* is not smart
